### DISCUSSION TITLE
Not able to configure MongoDb with Mongo Express username and password

### DISCUSSION THREAD

**Discussion by picBuilderLabs** [microsoft/aspire#5736] [Q&A] [answered]
I am trying to setup a mongodb database with a username and password and also trying to add  mongo express and protect it with a username and password.
- The error is when I set the password (Password12!) it says there are unescaped characters. so i fix the equivalent of the ! symbol and now i get the error: Could not connect to database using connectionString: mongodb://mongodb+srv://root:Password12%40mongodb:27017:27017/"

Here is my code sample I was playing around with
```
builder.AddMongoDB(name, port)
         .WithImage("mongo")
         .WithImageTag("7.0")
         .WithEnvironment("MONGO_INITDB_ROOT_USERNAME", "root")
         .WithEnvironment("MONGO_INITDB_ROOT_PASSWORD", "Password12!")
         //.WithDataVolume("mongovolume")
         .WithMongoExpress(config =>
         {
             config.WithHostPort(8091);
             config.WithEnvironment("ME_CONFIG_MONGODB_SERVER", "mongodb://root:Password12%40mongodb:27017"); // ! mark has to be unescaped or else it will fail
             config.WithEnvironment("ME_CONFIG_MONGODB_ENABLE_ADMIN", "true");
             config.WithEnvironment("ME_CONFIG_BASICAUTH_USERNAME", "admin");
             config.WithEnvironment("ME_CONFIG_BASICAUTH_PASSWORD", "Password12!");
             config.WithEnvironment("ME_CONFIG_SITE_BASEURL", "/mongo-express");
             config.WithEnvironment("ME_CONFIG_BASICAUTH", "true"); // Enable basic auth
         });
```

Any idea whats going on here?

**Comment by behdad088** [comment]
Hi @picBuilderLabs 
The issue you're encountering stems from the incorrect format and escaping of special characters in your MongoDB connection string. Let's break it down and resolve the issues:

### Issue 1: Unescaped Special Characters
In URLs (including MongoDB connection strings), special characters like `!`, `@`, `$`, etc., need to be URL-encoded to avoid being misinterpreted. In your case, the password `Password12!` includes the `!` symbol, which must be properly escaped as `%21`.

### Issue 2: Incorrect MongoDB Connection String Format
The connection string format you're using has an issue. MongoDB connection strings shouldn't contain both `mongodb://` and `mongodb+srv://` simultaneously, and the extra `:27017` is incorrect.

- `mongodb://`: This format is used for connecting to a standalone MongoDB instance or replica set.
- `mongodb+srv://`: This format is used for connecting to MongoDB Atlas or when using DNS-based seedlist discovery.

Given that you're trying to connect to a local MongoDB instance (port `27017`), you should use `mongodb://`.

### Fixes

1. **Escape special characters in the password**:
   - Escape the `!` character in the password by replacing it with `%21`.
   - Your password `Password12!` should become `Password12%21`.

2. **Correct the MongoDB connection string format**:
   - Fix the connection string to follow the proper format: `mongodb://root:Password12%21@mongodb:27017`.

### Final Code:

```csharp
builder.AddMongoDB(name, port)
       .WithImage("mongo")
       .WithImageTag("7.0")
       .WithEnvironment("MONGO_INITDB_ROOT_USERNAME", "root")
       .WithEnvironment("MONGO_INITDB_ROOT_PASSWORD", "Password12!")
       //.WithDataVolume("mongovolume")
       .WithMongoExpress(config =>
       {
           config.WithHostPort(8091);
           // Fixing connection string format and escaping the "!" character
           config.WithEnvironment("ME_CONFIG_MONGODB_SERVER", "mongodb://root:Password12%21@mongodb:27017");
           config.WithEnvironment("ME_CONFIG_MONGODB_ENABLE_ADMIN", "true");
           config.WithEnvironment("ME_CONFIG_BASICAUTH_USERNAME", "admin");
           // You should also escape the "!" character in the Mongo Express password
           config.WithEnvironment("ME_CONFIG_BASICAUTH_PASSWORD", "Password12%21");
           config.WithEnvironment("ME_CONFIG_SITE_BASEURL", "/mongo-express");
           config.WithEnvironment("ME_CONFIG_BASICAUTH", "true"); // Enable basic auth
       });
```

### Key Changes:
- **MongoDB connection string**: `mongodb://root:Password12%21@mongodb:27017`
- **Password escaping**: The `!` in the password is escaped as `%21`.

Now, the connection string should be properly formatted, and the special characters in the password will no longer cause issues. This should resolve your connection and authentication problems for both MongoDB and Mongo Express.

**Reply by picBuilderLabs** [reply, marked as answer]
Thank you for the reply, your post motivated me to get this darn thing resolved. These are the steps I took to resolve this issue:
- 'ME_CONFIG_MONGODB_SERVER' is deprecated and now 'ME_CONFIG_MONGODB_URL' is to be used instead. 
- After the @ symbol I had to use host.docker.internal, the image 'name' wouldn't get resolved
- ME_CONFIG_BASICAUTH_PASSWORD the password value needs NOT be unescaped.

Heres the code which works:
```
builder.AddMongoDB(name, port)
           .WithImage("mongo")
           .WithImageTag("7.0")
           .WithEnvironment("MONGO_INITDB_ROOT_USERNAME", "root")
           .WithEnvironment("MONGO_INITDB_ROOT_PASSWORD", "Password12!")
           //.WithDataVolume("mongovolume")
           .WithMongoExpress(config =>
           {
               config.WithHostPort(8091);
               // Fixing connection string format and escaping the "!" character
               config.WithEnvironment("ME_CONFIG_MONGODB_URL", "mongodb://root:Password12%21@host.docker.internal:27017");
               config.WithEnvironment("ME_CONFIG_MONGODB_ENABLE_ADMIN", "true");
               config.WithEnvironment("ME_CONFIG_BASICAUTH_USERNAME", "admin");
               config.WithEnvironment("ME_CONFIG_BASICAUTH_PASSWORD", "Password12!");
               config.WithEnvironment("ME_CONFIG_SITE_BASEURL", "/mongo-express");
               config.WithEnvironment("ME_CONFIG_BASICAUTH", "true"); // Enable basic auth
           });
```

the unfortunate part is that now even though the database itself is password protected and has a mongo express attached to it, on the client side the method 'builder.AddAspireMongoDb("mongo");' doesn't work cos the username and password is not passed here for the application to be able to do CRUD. Hence I had to change that to a native mongodb connection nullifying the positives I thought using AddAspireMongoDb would bring. Though it all flows and works, the code didn't stay native 'Aspire'. Kinda disappointed TBH.

**Reply by davidfowl** [reply]
PS you can see how username and password is being added to mongo here https://github.com/dotnet/aspire/pull/4370
